src/commands/dev/gcs/server.rs

```diff
@@ -0,0 +1,106 @@
+use crate::commands::dev::gcs::headers::{destructure_response, structure_request};
+use crate::commands::dev::server_config::ServerConfig;
+use crate::terminal::emoji;
+
+use std::sync::{Arc, Mutex};
+
+use chrono::prelude::*;
+use hyper::client::{HttpConnector, ResponseFuture};
+use hyper::header::{HeaderName, HeaderValue};
+use hyper::http::uri::InvalidUri;
+use hyper::service::{make_service_fn, service_fn};
+use hyper::{Body, Client as HyperClient, Request, Response, Server, Uri};
+use hyper_tls::HttpsConnector;
+
+const PREVIEW_HOST: &str = "rawhttp.cloudflareworkers.com";
+
+pub(super) async fn serve(
+    server_config: ServerConfig,
+    preview_id: Arc<Mutex<String>>,
+) -> Result<(), failure::Error> {
+    // set up https client to connect to the preview service
+    let https = HttpsConnector::new();
+    let client = HyperClient::builder().build::<_, Body>(https);
+
+    let listening_address = server_config.listening_address.clone();
+    // create a closure that hyper will use later to handle HTTP requests
+    let make_service = make_service_fn(move |_| {
+        let client = client.to_owned();
+        let preview_id = preview_id.lock().unwrap().to_owned();
+        let server_config = server_config.to_owned();
+        async move {
+            Ok::<_, failure::Error>(service_fn(move |req| {
+                let client = client.to_owned();
+                let preview_id = preview_id.to_owned();
+                let server_config = server_config.to_owned();
+                async move {
+                    let resp =
+                        preview_request(req, client, preview_id.to_owned(), server_config).await?;
+                    let (mut parts, body) = resp.into_parts();
+
+                    destructure_response(&mut parts)?;
+                    let resp = Response::from_parts(parts, body);
+                    Ok::<_, failure::Error>(resp)
+                }
+            }))
+        }
+    });
+
+    let server = Server::bind(&listening_address.address).serve(make_service);
+    println!("{} Listening on http://{}", emoji::EAR, listening_address);
+    if let Err(e) = server.await {
+        eprintln!("server error: {}", e);
```

Comment:
This means the local server errored right? since there are a few servers (localhost, previewer, the worker, ..) maybe you could specify what server errored?

Reply:
probably a good idea! not going to do that in this PR though as it was not introduced here (no new code in this PR so far, just reorganizing it)